syntax = "proto3";

option csharp_namespace = "Grpc.Terminal";

import "Contracts.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";

package terminal;

service MarketData {
  rpc GetSymbolTick (GetSymbolTickRequest) returns (GetSymbolTickReply) {}
  rpc StreamTicksRange (StreamTicksRangeRequest) returns (stream TicksRangeReply) {}
  rpc StreamTicksRangeBytes (StreamTicksRangeRequest) returns (stream TicksRangeBytesReply) {}
  rpc GetTicksRangeBytes (GetTicksRangeBytesRequest) returns (TicksRangeBytesReply) {}
  rpc StreamRatesRange (StreamRatesRangeRequest) returns (stream RatesRangeReply) {}
  rpc StreamRatesRangeFromTicks (StreamRatesRangeFromTicksRequest) returns (stream RatesRangeReply) {}
}

message GetSymbolTickRequest {
  string symbol = 1; 
}

message GetSymbolTickReply {
  Trade trade = 1;
  ResponseStatus responseStatus = 2;
}

message StreamTicksRangeRequest {
  string symbol = 1;
  google.protobuf.Timestamp fromDate = 2;
  google.protobuf.Timestamp toDate = 3;
  CopyTicks type = 4;
  int32 chunkSize = 5;
}

message TicksRangeReply {
  repeated Trade trades = 1;
  ResponseStatus responseStatus = 2;
}

message TicksRangeBytesReply {
  repeated int32 bytes = 1;
  ResponseStatus responseStatus = 2;
}

message GetTicksRangeBytesRequest {
  string symbol = 1;
  google.protobuf.Timestamp fromDate = 2;
  google.protobuf.Timestamp toDate = 3;
  CopyTicks type = 4;
}

message StreamRatesRangeRequest {
  string symbol = 1;
  google.protobuf.Timestamp fromDate = 2;
  google.protobuf.Timestamp toDate = 3;
  Timeframe timeframe = 4;
  int32 chunkSize = 5;
}

message RatesRangeReply {
  repeated Rate rates = 1;
  ResponseStatus responseStatus = 2;
}

message StreamRatesRangeFromTicksRequest {
  string symbol = 1;
  google.protobuf.Timestamp fromDate = 2;
  google.protobuf.Timestamp toDate = 3;
  google.protobuf.Duration timeframe = 4;
  int32 chunkSize = 5;
}

message Trade {
  google.protobuf.Timestamp time = 1;
  google.protobuf.DoubleValue bid = 2;
  google.protobuf.DoubleValue ask = 3;
  google.protobuf.DoubleValue last = 4;
  google.protobuf.DoubleValue volume = 5;
  TickFlags flags = 6;
  google.protobuf.DoubleValue volumeReal = 7;
}

message Rate {
  google.protobuf.Timestamp time = 1;
  google.protobuf.DoubleValue open = 2;
  google.protobuf.DoubleValue high = 3;
  google.protobuf.DoubleValue low = 4;
  google.protobuf.DoubleValue close = 5;
  google.protobuf.DoubleValue tickVolume = 6;
  google.protobuf.DoubleValue spread = 7;
  google.protobuf.DoubleValue volume = 8;
}
